LanguageExt.Core/DataTypes/Seq/Enum.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace LanguageExt
+{
+    internal class Enum<A>
+    {
+        IEnumerator<A> iter;
+        List<A> list;
+
+        public Enum(IEnumerable<A> seq)
+        {
+            this.iter = seq.GetEnumerator();
+            this.list = new List<A>();
+        }
+
+        public Enum(IEnumerator<A> iter)
+        {
+            this.iter = iter;
+            this.list = new List<A>();
+        }
+
+        ~Enum()
+        {
+            iter?.Dispose();
+            iter = null;
+        }
+
+        public (bool Success, A Value) Get(int index)
+        {
+            if (index < list.Count)
+            {
+                return (true, list[index]);
+            }
+            if (iter == null)
+            {
+                return (false, default(A));
+            }
+            else
+            {
+                lock (list)
+                {
+                    bool theresMore = true;
+                    while (index >= list.Count && (theresMore = iter.MoveNext()))
+                    {
+                        list.Add(iter.Current);
+                    }
+                    if (!theresMore)
+                    {
+                        iter = null;
+                    }
+                }
+            }
+            if (index < list.Count)
```

Comment:
Since this is common with the earlier code in the get, maybe we should check and grow, then fall through?

Reply:
Sorry, I don't follow?  I'm looking at it now, and I can see that last `if` should be moved inside the `else` above it.  But other than that not really sure.  The ugliness of the function is obviously to try and return as quickly as possible without doing unnecessary work.